stan/math/prim/mat/fun/autocorrelation.hpp

```diff
@@ -103,6 +104,53 @@ void autocorrelation(const std::vector<T>& y, std::vector<T>& ac,
     ac[i] /= var;
 }
 
+/**
+ * Write autocorrelation estimates for every lag for the specified
+ * input sequence into the specified result using the specified
+ * FFT engine.  The return vector be resized to the same length as
+ * the input sequence with lags given by array index.
+ *
+ * <p>The implementation involves a fast Fourier transform,
+ * followed by a normalization, followed by an inverse transform.
+ *
+ * <p>An FFT engine can be created for reuse for type double with:
+ *
+ * <pre>
+ *     Eigen::FFT<double> fft;
+ * </pre>
+ *
+ * @tparam T Scalar type.
+ * @param y Input sequence.
+ * @param ac Autocorrelations.
+ * @param fft FFT engine instance.
+ */
+template <typename T, typename DerivedA, typename DerivedB>
+void autocorrelation(const Eigen::MatrixBase<DerivedA>& y,
```

Comment:
[optional]
I'd be inclined to remove the old implementation for `std::vector` and instead call the new Eigen implementation with an `Eigen::Map` for the first two arguments around `&y[0]` and `&ac[0]` (the latter after resizing), a reference to the first element of `y` and `ac`.

Reply:
I like this idea, but we need to coordinate with the stan-dev/stan repository as there is some code there that still uses the `std::vector` signature.  Once this PR is in, we can clean up some other stuff in stan/src/stan/mcmc/chains.hpp which will then enable us to come back to this.

Reply:
I didn't mean to suggest we'd remove the signature, just change the implementation to call the Eigen implementation rather than implementing everything twice.

Reply:
Great thought.  Let me give that a go.